File: c++/single-number-ii.cpp
class Solution {
public:
    int singleNumber(int A[], int n) {
        /* https://oj.leetcode.com/problems/single-number-ii/
        count[0..31]: Bookkeeping the number of occurences of every bit.
        (for a int, totally 32 bits)
        If count[i] % 3 != 0, then the ith bit beblongs to the single number.
        */
        
        int count[32];
        memset(count, 0, sizeof(int)*32);
        
        for (int i = 0; i < n; i++) {
             for (int j = 0; j < 32; j++) {
                 if ((A[i] >> j) & 1) {
                     count[j]++;
                 }
             }
        }
        
        int result = 0;
        for (int i = 0; i < 32; i++) {
            if (count[i] % 3) result |= (1 << i);
        }
        
        return result;
    }
};

File: python/single-number-ii.py
class Solution:
    # @param A, a list of integer
    # @return an integer
    def singleNumber(self, A):
        bit_counts = [0] * 32
        neg_count = 0

        for x in A:
            if x < 0:
                neg_count += 1
                x = -x
            for i in xrange(32):
                if (x >> i) & 1:
                    bit_counts[i] += 1

        result = 0
        for i, bc in enumerate(bit_counts):
            if bc % 3:
                result |= 1 << i

        if neg_count % 3:
            result = -result

        return result
